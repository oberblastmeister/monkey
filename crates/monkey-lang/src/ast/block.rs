use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub open_bracket: T!['{'],
    pub stmts: Vec<ast::Stmt>,
    pub close_bracket: T!['}'],
}

impl Parse for Block {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        let open_bracket = p.parse()?;

        let mut stmts = Vec::new();

        loop {
            stmts.push(p.parse()?);
            if p.is_eof()? {
                break;
            }
        }

        let close_bracket = p.parse()?;

        Ok(Block {
            open_bracket,
            stmts,
            close_bracket,
        })
    }
}
