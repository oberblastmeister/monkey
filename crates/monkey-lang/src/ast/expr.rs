use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Binary(Box<ast::ExprBinary>),
    Lit(ast::Lit),
}

impl Parse for Expr {
    #[inline]
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        expr(p)
    }
}

#[inline]
fn expr(p: &mut Parser) -> ParseResult<Expr> {
    expr_bp(p, 0)
}

fn expr_bp(p: &mut Parser, min_bp: u8) -> ParseResult<Expr> {
    let mut lhs = match p.nth(0)? {
        K!['('] => expr_paren(p)?,
        _ if p.peek::<ast::Lit>() => Expr::Lit(p.parse()?),
        t => panic!("Bad token: {:?}", t),
    };

    loop {
        if p.peek::<ExprEnd>() {
            break;
        };

        let op = ast::BinOp::from_peeker(&mut p.peeker).ok_or_else(|| {
            ParseError::expected(&p.next().unwrap(), "Expected a binary operator")
        })?;

        let (l_bp, r_bp) = op.binding_power();

        if l_bp < min_bp {
            break;
        }

        lhs = {
            let token = p.next()?;

            let rhs = expr_bp(p, r_bp)?;

            Expr::Binary(Box::new(ast::ExprBinary {
                lhs,
                op,
                rhs,
                token,
            }))
        };

        continue;
    }

    Ok(lhs)
}

struct ExprEnd;

impl Peek for ExprEnd {
    fn peek(p: &mut Peeker) -> bool {
        matches!(p.nth(0), K![eof] | K![;] | K![')'])
    }
}

fn expr_paren(p: &mut Parser) -> ParseResult<Expr> {
    assert_eq!(p.nth(0).unwrap(), K!['(']);

    p.next().unwrap();
    let lhs = expr_bp(p, 0)?;
    p.parse::<T![')']>()?;
    Ok(lhs)
}
