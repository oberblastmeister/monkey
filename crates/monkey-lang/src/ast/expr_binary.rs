use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprBinary {
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    pub op: BinOp,
    pub rhs: ast::Expr,
}

impl ExprBinary {
    pub fn with_lhs(
        p: &mut Parser<'_>,
        lhs: ast::Expr,
        op: BinOp,
    ) -> ParseResult<ExprBinary> {
        assert!(BinOp::from_peeker(&mut p.peeker).is_some());

        let token = p.next()?;

        Ok(ExprBinary {
            lhs,
            token,
            op,
            rhs: p.parse()?,
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
}

use BinOp::*;

impl BinOp {
    pub fn from_peeker(p: &mut Peeker<'_>) -> Option<BinOp> {
        Some(match p.nth(0) {
            K![+] => Add,
            K![-] => Sub,
            K![*] => Mul,
            K![/] => Div,
            _ => return None,
        })
    }

    pub fn binding_power(self) -> (u8, u8) {
        match self {
            Add | Sub => (1, 2),
            Mul | Div => (3, 4),
        }
    }
}

impl Peek for BinOp {
    fn peek(p: &mut Peeker) -> bool {
        BinOp::from_peeker(p).is_some()
    }
}
