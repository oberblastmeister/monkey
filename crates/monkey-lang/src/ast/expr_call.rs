use crate::{ast, Parse, ParseResult, Parser};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprCall {
    ident: T![ident],
    arguments: Arguments,
}

impl Parse for ExprCall {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        Ok(ExprCall {
            ident: p.parse()?,
            arguments: p.parse()?,
        })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Arguments(ast::List<T![ident], T!['{'], T!['}']>);

impl Parse for Arguments {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        Ok(Arguments(p.parse()?))
    }
}
