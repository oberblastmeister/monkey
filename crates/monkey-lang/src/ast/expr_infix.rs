use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprInfix {
    /// The left hand side of the expression
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
    /// The right hand side of the expression
    pub rhs: ast::Expr,
}

expr_parse!(Infix, ExprInfix, "binary expression");
