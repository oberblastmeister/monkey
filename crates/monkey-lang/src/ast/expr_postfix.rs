use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprPostfix {
    /// The right hand side of the expression
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
}

expr_parse!(Postfix, ExprPostfix, "postfix expression");
