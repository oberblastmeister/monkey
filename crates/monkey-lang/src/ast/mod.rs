#[macro_use]
pub mod generated;

pub use generated::TokenKind;

mod token;
mod expr;
mod expr_lit;
mod expr_binary;
mod lit;
mod lit_bool;
mod lit_num;
mod lit_str;
mod stmt;
mod file;
mod op;

pub use token::Token;

pub use expr::Expr;
pub use expr_lit::ExprLit;
pub use expr_binary::{BinOp, ExprBinary};

pub use lit::Lit;
pub use lit_bool::LitBool;
pub use lit_num::LitNum;
pub use lit_str::LitStr;

pub use stmt::Stmt;
pub use file::File;
