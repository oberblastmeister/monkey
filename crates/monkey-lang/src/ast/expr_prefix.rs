use crate::ast;

#[derive(Debug, Clone, PartialEq)]
pub struct ExprPrefix {
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
    /// The right hand side of the expression
    pub rhs: ast::Expr,
}

expr_parse!(Prefix, ExprPrefix, "prefix expression");
