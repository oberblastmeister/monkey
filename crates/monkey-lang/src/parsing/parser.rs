use std::collections::VecDeque;

use super::{Lexer, Parse, ParseError, ParseErrorKind, ParseResult, Peek};
use crate::ast::{self, Token, TokenKind};
use crate::Span;

/// Construct used to peek a parser.
#[derive(Debug)]
pub struct Peeker<'a> {
    pub(crate) lexer: Lexer<'a>,

    buf: VecDeque<Token>,

    last: Option<Span>,
}

impl<'a> Peeker<'a> {
    pub fn new(input: &'a str) -> Peeker<'a> {
        let lexer = Lexer::new(input);
        let buf = VecDeque::new();
        let last = None;
        Peeker { lexer, buf, last }
    }

    /// Peek the token kind at the given position.
    pub fn nth(&mut self, n: usize) -> TokenKind {
        match self.at(n) {
            Ok(t) => match t {
                Some(t) => t.kind,
                None => TokenKind::Eof,
            },
            Err(error) => TokenKind::Error,
        }
    }

    /// Make sure there are at least `n` items in the buffer, and return the
    /// item at that point.
    fn at(&mut self, n: usize) -> ParseResult<Option<&Token>> {
        while self.buf.len() <= n {
            let token = match self.lexer.next_token() {
                Ok(token) => token,
                Err(e) if e.kind() == &ParseErrorKind::Eof => break,
                Err(e) => return Err(e),
            };

            self.last = Some(token.span);
            self.buf.push_back(token);
        }

        Ok(self.buf.get(n))
    }

    /// Peek something else
    pub fn peek<T>(&mut self) -> bool
    where
        T: Peek,
    {
        T::peek(self)
    }
}

#[derive(Debug)]
pub struct Parser<'a> {
    pub peeker: Peeker<'a>,
    errors: Vec<ParseError>,
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        Parser {
            peeker: Peeker::new(input),
            errors: Vec::new(),
        }
    }

    /// Parse a specific item from the parser.
    pub fn parse<T>(&mut self) -> ParseResult<T>
    where
        T: Parse,
    {
        T::parse(self)
    }

    pub fn peek<T>(&mut self) -> bool
    where
        T: Peek,
    {
        T::peek(&mut self.peeker)
    }

    /// Try to consume a single thing matching `T`, returns `true` if any tokens
    /// were consumed.
    pub fn try_parse<T>(&mut self) -> ParseResult<bool>
    where
        T: Parse + Peek,
    {
        Ok(if self.peek::<T>() {
            self.parse::<T>()?;
            true
        } else {
            false
        })
    }

    /// Try to consume all things matching `T`, returns `true` if any tokens
    /// were consumed.
    pub fn try_parse_all<T>(&mut self) -> ParseResult<bool>
    where
        T: Parse + Peek,
    {
        let mut consumed = false;

        while self.peek::<T>() {
            self.parse::<T>()?;
            consumed = true;
        }

        Ok(consumed)
    }

    /// Consume the next token from the parser.
    #[allow(clippy::should_implement_trait)]
    pub fn next(&mut self) -> ParseResult<Token> {
        if let Some(t) = self.peeker.buf.pop_front() {
            return Ok(t);
        }

        match self.peeker.lexer.next_token() {
            Err(e) if e.kind() == &ParseErrorKind::Eof => Ok(Token {
                span: Span::new(0, 0),
                kind: TokenKind::Eof,
                text: "".into(),
            }),
            Err(e) => Err(e),
            Ok(t) => Ok(t),
        }
    }

    /// Peek the token kind at the given position.
    pub fn nth(&mut self, n: usize) -> ParseResult<TokenKind> {
        if let Some(t) = self.peeker.at(n)? {
            Ok(t.kind)
        } else {
            Ok(TokenKind::Eof)
        }
    }

    /// Test if the parser is at end-of-file, after which there is no more input
    /// to parse.
    pub fn is_eof(&mut self) -> Result<bool, ParseError> {
        Ok(self.peeker.at(0)?.is_none())
    }

    /// Assert that the parser has reached its end-of-file.
    pub fn eof(&mut self) -> Result<(), ParseError> {
        if let Some(token) = self.peeker.at(0)? {
            return Err(ParseError::new(
                token,
                ParseErrorKind::ExpectedEof {
                    actual: token.kind.as_str(),
                },
            ));
        }

        Ok(())
    }
}

pub fn parse<T: Parse>(input: &str) -> ParseResult<T> {
    Parser::new(input).parse::<T>()
}

#[cfg(test)]
mod tests {
    use std::fs;

    use super::*;

    #[test]
    fn parser() {
        insta::glob!("snapshot_inputs/parser/*.txt", |path| {
            let input = fs::read_to_string(path).unwrap();
            let suffix = path.file_stem().unwrap().to_str().unwrap();
            insta::with_settings!({snapshot_path => "snapshots/parser", snapshot_suffix => suffix}, {
                insta::assert_debug_snapshot!(parse::<ast::File>(&input).map(|f| f.stmts))
            })
        })
    }
}
