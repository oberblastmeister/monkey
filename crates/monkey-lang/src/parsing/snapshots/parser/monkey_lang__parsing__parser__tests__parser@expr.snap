---
source: crates/monkey-lang/src/parsing/parser.rs
expression: "parse::<ast::File>(&input).map(|f| f.stmts)"

---
Ok(
    [
        Expr(
            Binary(
                ExprBinary {
                    lhs: Binary(
                        ExprBinary {
                            lhs: Binary(
                                ExprBinary {
                                    lhs: Lit(
                                        Num(
                                            LitNum {
                                                num: Number {
                                                    token: Token {
                                                        span: Span {
                                                            start: 0,
                                                            end: 1,
                                                        },
                                                        kind: Number,
                                                        text: "1",
                                                    },
                                                },
                                                value: 1.0,
                                            },
                                        ),
                                    ),
                                    token: Token {
                                        span: Span {
                                            start: 2,
                                            end: 3,
                                        },
                                        kind: Plus,
                                        text: "+",
                                    },
                                    op: Add,
                                    rhs: Lit(
                                        Num(
                                            LitNum {
                                                num: Number {
                                                    token: Token {
                                                        span: Span {
                                                            start: 4,
                                                            end: 5,
                                                        },
                                                        kind: Number,
                                                        text: "1",
                                                    },
                                                },
                                                value: 1.0,
                                            },
                                        ),
                                    ),
                                },
                            ),
                            token: Token {
                                span: Span {
                                    start: 6,
                                    end: 7,
                                },
                                kind: Plus,
                                text: "+",
                            },
                            op: Add,
                            rhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 8,
                                                    end: 9,
                                                },
                                                kind: Number,
                                                text: "1",
                                            },
                                        },
                                        value: 1.0,
                                    },
                                ),
                            ),
                        },
                    ),
                    token: Token {
                        span: Span {
                            start: 10,
                            end: 11,
                        },
                        kind: Plus,
                        text: "+",
                    },
                    op: Add,
                    rhs: Lit(
                        Num(
                            LitNum {
                                num: Number {
                                    token: Token {
                                        span: Span {
                                            start: 12,
                                            end: 13,
                                        },
                                        kind: Number,
                                        text: "1",
                                    },
                                },
                                value: 1.0,
                            },
                        ),
                    ),
                },
            ),
            Semicolon {
                token: Token {
                    span: Span {
                        start: 13,
                        end: 14,
                    },
                    kind: Semicolon,
                    text: ";",
                },
            },
        ),
        Expr(
            Binary(
                ExprBinary {
                    lhs: Binary(
                        ExprBinary {
                            lhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 15,
                                                    end: 16,
                                                },
                                                kind: Number,
                                                text: "1",
                                            },
                                        },
                                        value: 1.0,
                                    },
                                ),
                            ),
                            token: Token {
                                span: Span {
                                    start: 17,
                                    end: 18,
                                },
                                kind: Minus,
                                text: "-",
                            },
                            op: Sub,
                            rhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 19,
                                                    end: 23,
                                                },
                                                kind: Number,
                                                text: "3134",
                                            },
                                        },
                                        value: 3134.0,
                                    },
                                ),
                            ),
                        },
                    ),
                    token: Token {
                        span: Span {
                            start: 24,
                            end: 25,
                        },
                        kind: Plus,
                        text: "+",
                    },
                    op: Add,
                    rhs: Lit(
                        Num(
                            LitNum {
                                num: Number {
                                    token: Token {
                                        span: Span {
                                            start: 26,
                                            end: 29,
                                        },
                                        kind: Number,
                                        text: "134",
                                    },
                                },
                                value: 134.0,
                            },
                        ),
                    ),
                },
            ),
            Semicolon {
                token: Token {
                    span: Span {
                        start: 29,
                        end: 30,
                    },
                    kind: Semicolon,
                    text: ";",
                },
            },
        ),
        Expr(
            Binary(
                ExprBinary {
                    lhs: Binary(
                        ExprBinary {
                            lhs: Binary(
                                ExprBinary {
                                    lhs: Lit(
                                        Num(
                                            LitNum {
                                                num: Number {
                                                    token: Token {
                                                        span: Span {
                                                            start: 31,
                                                            end: 34,
                                                        },
                                                        kind: Number,
                                                        text: "123",
                                                    },
                                                },
                                                value: 123.0,
                                            },
                                        ),
                                    ),
                                    token: Token {
                                        span: Span {
                                            start: 35,
                                            end: 36,
                                        },
                                        kind: Asterisk,
                                        text: "*",
                                    },
                                    op: Mul,
                                    rhs: Lit(
                                        Num(
                                            LitNum {
                                                num: Number {
                                                    token: Token {
                                                        span: Span {
                                                            start: 37,
                                                            end: 41,
                                                        },
                                                        kind: Number,
                                                        text: "2341",
                                                    },
                                                },
                                                value: 2341.0,
                                            },
                                        ),
                                    ),
                                },
                            ),
                            token: Token {
                                span: Span {
                                    start: 42,
                                    end: 43,
                                },
                                kind: Minus,
                                text: "-",
                            },
                            op: Sub,
                            rhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 44,
                                                    end: 47,
                                                },
                                                kind: Number,
                                                text: "234",
                                            },
                                        },
                                        value: 234.0,
                                    },
                                ),
                            ),
                        },
                    ),
                    token: Token {
                        span: Span {
                            start: 48,
                            end: 49,
                        },
                        kind: Plus,
                        text: "+",
                    },
                    op: Add,
                    rhs: Lit(
                        Num(
                            LitNum {
                                num: Number {
                                    token: Token {
                                        span: Span {
                                            start: 50,
                                            end: 53,
                                        },
                                        kind: Number,
                                        text: "134",
                                    },
                                },
                                value: 134.0,
                            },
                        ),
                    ),
                },
            ),
            Semicolon {
                token: Token {
                    span: Span {
                        start: 53,
                        end: 54,
                    },
                    kind: Semicolon,
                    text: ";",
                },
            },
        ),
        Expr(
            Binary(
                ExprBinary {
                    lhs: Lit(
                        Num(
                            LitNum {
                                num: Number {
                                    token: Token {
                                        span: Span {
                                            start: 55,
                                            end: 57,
                                        },
                                        kind: Number,
                                        text: "34",
                                    },
                                },
                                value: 34.0,
                            },
                        ),
                    ),
                    token: Token {
                        span: Span {
                            start: 58,
                            end: 59,
                        },
                        kind: Plus,
                        text: "+",
                    },
                    op: Add,
                    rhs: Binary(
                        ExprBinary {
                            lhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 60,
                                                    end: 64,
                                                },
                                                kind: Number,
                                                text: "1324",
                                            },
                                        },
                                        value: 1324.0,
                                    },
                                ),
                            ),
                            token: Token {
                                span: Span {
                                    start: 65,
                                    end: 66,
                                },
                                kind: Asterisk,
                                text: "*",
                            },
                            op: Mul,
                            rhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 67,
                                                    end: 70,
                                                },
                                                kind: Number,
                                                text: "443",
                                            },
                                        },
                                        value: 443.0,
                                    },
                                ),
                            ),
                        },
                    ),
                },
            ),
            Semicolon {
                token: Token {
                    span: Span {
                        start: 70,
                        end: 71,
                    },
                    kind: Semicolon,
                    text: ";",
                },
            },
        ),
        Expr(
            Binary(
                ExprBinary {
                    lhs: Lit(
                        Num(
                            LitNum {
                                num: Number {
                                    token: Token {
                                        span: Span {
                                            start: 72,
                                            end: 74,
                                        },
                                        kind: Number,
                                        text: "34",
                                    },
                                },
                                value: 34.0,
                            },
                        ),
                    ),
                    token: Token {
                        span: Span {
                            start: 75,
                            end: 76,
                        },
                        kind: Asterisk,
                        text: "*",
                    },
                    op: Mul,
                    rhs: Binary(
                        ExprBinary {
                            lhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 78,
                                                    end: 80,
                                                },
                                                kind: Number,
                                                text: "48",
                                            },
                                        },
                                        value: 48.0,
                                    },
                                ),
                            ),
                            token: Token {
                                span: Span {
                                    start: 81,
                                    end: 82,
                                },
                                kind: Plus,
                                text: "+",
                            },
                            op: Add,
                            rhs: Lit(
                                Num(
                                    LitNum {
                                        num: Number {
                                            token: Token {
                                                span: Span {
                                                    start: 83,
                                                    end: 84,
                                                },
                                                kind: Number,
                                                text: "3",
                                            },
                                        },
                                        value: 3.0,
                                    },
                                ),
                            ),
                        },
                    ),
                },
            ),
            Semicolon {
                token: Token {
                    span: Span {
                        start: 85,
                        end: 86,
                    },
                    kind: Semicolon,
                    text: ";",
                },
            },
        ),
    ],
)
